crates/ripemd160/src/ripemd160.rs

```diff
@@ -0,0 +1,408 @@
+//! Circuits for the [RIPEMD-160] hash function and its internal compression
+//! function.
+//!
+//! [RIPEMD-160]: https://homes.esat.kuleuven.be/~bosselae/ripemd160.html
+
+use bellpepper::gadgets::{multieq::MultiEq, uint32::UInt32};
+use bellpepper_core::{boolean::Boolean, ConstraintSystem, SynthesisError};
+use ff::PrimeField;
+use std::convert::TryInto;
+
+use crate::util::{f1, f2, f3, f4, f5, swap_byte_endianness, uint32_rotl};
+
+const IV: [u32; 5] = [0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0];
+const ROUND_CONSTANTS_LEFT: [u32; 5] = [0x00000000, 0x5a827999, 0x6ed9eba1, 0x8f1bbcdc, 0xa953fd4e];
+const ROUND_CONSTANTS_RIGHT: [u32; 5] =
+    [0x50a28be6, 0x5c4dd124, 0x6d703ef3, 0x7a6d76e9, 0x00000000];
+
+pub const MSG_SEL_IDX_LEFT: [usize; 80] = [
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 7, 4, 13, 1, 10, 6, 15, 3, 12, 0, 9, 5,
+    2, 14, 11, 8, 3, 10, 14, 4, 9, 15, 8, 1, 2, 7, 0, 6, 13, 11, 5, 12, 1, 9, 11, 10, 0, 8, 12, 4,
+    13, 3, 7, 15, 14, 5, 6, 2, 4, 0, 5, 9, 7, 12, 2, 10, 14, 1, 3, 8, 11, 6, 15, 13,
+];
+pub const MSG_SEL_IDX_RIGHT: [usize; 80] = [
+    5, 14, 7, 0, 9, 2, 11, 4, 13, 6, 15, 8, 1, 10, 3, 12, 6, 11, 3, 7, 0, 13, 5, 10, 14, 15, 8, 12,
+    4, 9, 1, 2, 15, 5, 1, 3, 7, 14, 6, 9, 11, 8, 12, 2, 10, 0, 4, 13, 8, 6, 4, 1, 3, 11, 15, 0, 5,
+    12, 2, 13, 9, 7, 10, 14, 12, 15, 10, 4, 1, 5, 8, 7, 6, 2, 13, 14, 0, 3, 9, 11,
+];
+
+pub const ROL_AMOUNT_LEFT: [usize; 80] = [
+    11, 14, 15, 12, 5, 8, 7, 9, 11, 13, 14, 15, 6, 7, 9, 8, 7, 6, 8, 13, 11, 9, 7, 15, 7, 12, 15,
+    9, 11, 7, 13, 12, 11, 13, 6, 7, 14, 9, 13, 15, 14, 8, 13, 6, 5, 12, 7, 5, 11, 12, 14, 15, 14,
+    15, 9, 8, 9, 14, 5, 6, 8, 6, 5, 12, 9, 15, 5, 11, 6, 8, 13, 12, 5, 12, 13, 14, 11, 8, 5, 6,
+];
+pub const ROL_AMOUNT_RIGHT: [usize; 80] = [
+    8, 9, 9, 11, 13, 15, 15, 5, 7, 7, 8, 11, 14, 14, 12, 6, 9, 13, 15, 7, 12, 8, 9, 11, 7, 7, 12,
+    7, 6, 15, 13, 11, 9, 7, 15, 11, 8, 6, 6, 14, 12, 13, 5, 14, 13, 13, 7, 5, 15, 5, 8, 11, 14, 14,
+    6, 14, 6, 9, 12, 9, 12, 5, 15, 8, 8, 5, 12, 9, 12, 5, 14, 6, 8, 13, 6, 5, 15, 13, 11, 11,
+];
+
+pub fn ripemd160<Scalar, CS>(cs: CS, input: &[Boolean]) -> Result<[Boolean; 160], SynthesisError>
+where
+    Scalar: PrimeField,
+    CS: ConstraintSystem<Scalar>,
+{
+    assert!(input.len() % 8 == 0);
+
+    let mut padded = input.to_vec();
+    let plen = padded.len() as u64;
+    padded.push(Boolean::Constant(true));
+
+    while (padded.len() + 64) % 512 != 0 {
```

Comment:
Not super important, but you can probably compute how the size of the final vector in advance and allocate that size earlier, which avoids stressing the allocator with adding elements one-by-one:
```
let zero_bytes = 512 - (padded.len() + 64) % 512;
padded.extend(iter::repeat(Boolean::Constant(false)).take(zero_bytes));
```

Reply:
Thanks for the suggestion. The while loop idea is from the original [sha256](https://github.com/lurk-lab/bellpepper/blob/34b537995d78de1aa1b0ee7c904fa163399dad1f/crates/bellpepper/src/gadgets/sha256.rs#L62) gadet. I have made a slight change to your suggestion. If the message after appending the single 1 bit and the 64-bit message length has a length which is a multiple of 512, we don't need to append zero bits.

```rust
    let num_zero_bits = 512 - (padded.len() + 64) % 512;
    // If num_zero_bits == 512, then (padded.len() + 64) is already a multiple of 512
    // No zero bits need to be appended in this case
    if num_zero_bits != 512 {
        padded.extend(iter::repeat(Boolean::Constant(false)).take(num_zero_bits));
    }
```